bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,551 @@
+      private ResultScanner scanner;
+      private Table table;
+      private Result current;
+
+      private Reader(
+          BoundedSource<Result> source, CloudBigtableScanConfiguration config, Scan scan) {
+        this.source = source;
+        this.config = config;
+        this.scan = scan;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        current = scanner.next();
+        return current != null;
+      }
+
+      @Override
+      public void close() throws IOException {
+        scanner.close();
+        table.close();
+        connection.close();
+      }
+
+      @Override
+      public Result getCurrent() throws NoSuchElementException {
+        return current;
+      }
+
+      @Override
+      public BoundedSource<Result> getCurrentSource() {
+        return source;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        connection = new BigtableConnection(config.toHBaseConfig());
+        table = connection.getTable(TableName.valueOf(config.getTableId()));
+        scanner = table.getScanner(scan);
+        return advance();
+      }
+    }
+
+    public Source(CloudBigtableScanConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public Coder<Result> getDefaultOutputCoder() {
+      return new HBaseResultCoder();
+    }
+
+    /**
+     * Splits the table based on keys that belong to "regions" a.k.a. tablets via the HBase
+     * {@link RegionLocator} interface which calls
+     * {@link BigtableService#sampleRowKeys(com.google.bigtable.v1.SampleRowKeysRequest, io.grpc.stub.StreamObserver)
+     * under the covers}.  Each BoundedResource will correspond to a single region.
+     */
+    @Override
+    public List<SourceWithKeys> splitIntoBundles(
+        long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+      Scan scan = configuration.getScan();
+      byte[] scanStartKey = scan.getStartRow();
+
+      byte[] scanEndKey = scan.getStopRow();
+      List<SourceWithKeys> splits = new ArrayList<>();
+      BigtableOptions bigtableOptions = configuration.toBigtableOptions();
+      try (BigtableSession session = new BigtableSession(bigtableOptions)) {
+        BigtableClusterName clusterName = bigtableOptions.getClusterName();
+        BigtableTableName tableName = clusterName.toTableName(configuration.getTableId());
+        SampleRowKeysRequest.Builder request = SampleRowKeysRequest.newBuilder();
+        request.setTableName(tableName.toString());
+
+        List<SampleRowKeysResponse> sampleRowKeys =
+            session.getDataClient().sampleRowKeys(request.build());
+        byte[] startKey = HConstants.EMPTY_START_ROW;
+        long lastOffset = 0;
+        for (SampleRowKeysResponse response : sampleRowKeys) {
+          byte[] endKey = response.getRowKey().toByteArray();
+          // Avoid empty regions.
+          if (Bytes.equals(startKey, endKey) && startKey.length > 0) {
```

Comment:
Is this possible? also, whystartKey.length > 0? Maybe ==HConstants.EMPTY_START_ROW?